pkg/kubelet/lifecycle/predicate.go

```diff
@@ -247,5 +250,16 @@ func GeneralPredicates(pod *v1.Pod, nodeInfo *schedulerframework.NodeInfo) ([]Pr
 		reasons = append(reasons, &PredicateFailureError{nodeports.Name, nodeports.ErrReason})
 	}
 
+	// Check taint/toleration except for static pods
+	if !types.IsStaticPod(pod) {
+		_, isUntolerated := corev1.FindMatchingUntoleratedTaint(nodeInfo.Node().Spec.Taints, pod.Spec.Tolerations, func(t *v1.Taint) bool {
+			// PodToleratesNodeTaints is only interested in NoSchedule and NoExecute taints.
+			return t.Effect == v1.TaintEffectNoSchedule || t.Effect == v1.TaintEffectNoExecute
```

Comment:
NoSchedule shouldn't be considered by kubelet, only NoExecute, right?

Reply:
also, if Effect is unset, this should tolerate all effects, right?

Reply:
`NoSchedule` means that a Pod is not scheduled or started in a node.
`NoExecute` means that a Pod is not scheduled or started in a node and that it's evicted if it's already in the node.

Then, both indicate that the Pod shouldn't be started. `NoExecute` has the additional effect of affecting running Pods.

AFAIK, taints always have an effect (tolerations might not).

Reply:
> NoSchedule means that a Pod is not scheduled or started in a node.

no, it explicitly means it's not scheduled to a node, but nodes submitted directly to the node without the scheduler are permitted:

```
	// Do not allow new pods to schedule onto the node unless they tolerate the taint,
	// but allow all pods submitted to Kubelet without going through the scheduler
	// to start, and allow all already-running pods to continue running.
	// Enforced by the scheduler.
	TaintEffectNoSchedule TaintEffect = "NoSchedule"
```

also, I think by rejecting the pod here, an already-running pod would be stopped if the kubelet restarted

Reply:
Ah, it doesn't match the user documentation :/ https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/ (at least it's not clear in this regard).

But since the API comment is clear, let's just check NoSchedule.

Reply:
> But since the API comment is clear, let's just check NoSchedule.

NoExecute?

Reply:
Oops, yes

Reply:
That seems more reasonable to me, but would like a second from the node team.

Additionally, since this break passed tests as-is, we have a test gap that would allow breaking handling of NoSchedule taints. Please add a test that would have caught this to ensure the kubelet does not prevent admission of pods for that reason.